### PR TITLE
Add uint128 and uint256 for smart contracts

### PR DESCRIPTION
**Pull request by quantumagi**

### CODE REVIEW

**Comment by quantumagi** [comment]
Ready for review.

**Comment by quantumagi** [comment]
Ready for review.

**Comment by quantumagi** [comment]
Ready for review

**Comment by quantumagi** [comment]
See https://github.com/stratisproject/StratisFullNode/pull/365.

File: Stratis.SmartContracts.Tests/UInt256Tests.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            UInt256 x = 50;
+            Assert.Equal((ulong)50, (ulong)x);
+        }
+
+        [Fact]
+        [Trait("UnitTest", "UnitTest")]
+        public void Uint256FromTooLargeUlongThrowsError()
+        {
+            UInt256 x = UInt256.Parse("010000000000000000");
+            Assert.Throws(typeof(OverflowException), () => (ulong)x);
```
Comment (YakupIpek) [inline Stratis.SmartContracts.Tests/UInt256Tests.cs]: XUnit gives warning that "Don't use typeof expression to check the exception type" and recommends
`Assert.Throws<OverflowException>(() => (ulong)x)`

File: Stratis.SmartContracts.Tests/UInt256Tests.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using Xunit;
+
+namespace Stratis.SmartContracts.Tests
+{
+    public class UInt256Tests
+    {
+        [Fact]
+        [Trait("UnitTest", "UnitTest")]
```
Comment (YakupIpek) [inline Stratis.SmartContracts.Tests/UInt256Tests.cs]: Trait attribute is needed ? The other tests don't use it and it is nice to keep it consistent

File: Stratis.SmartContracts.Tests/UInt256Tests.cs (hunk 1 above)
Comment (YakupIpek) [reply Stratis.SmartContracts.Tests/UInt256Tests.cs]: It is expected behavior for numeric types that it shouldn't throw overflow exception when there is overflow as default behavior. Only it should under checked context. 

I see that this is default behavior of `BigInteger` which is internal of `UInt128` and `UInt256` so i think it is fine

File: Stratis.SmartContracts/UIntBase.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        protected BigInteger ShiftRight(int shift) => this.value >> shift;
+
+        protected BigInteger ShiftLeft(int shift) => this.value << shift;
+
+        protected BigInteger Add(BigInteger value2) => this.value + value2;
+
+        protected BigInteger Subtract(BigInteger value2)
+        {
+            if (this.value.CompareTo(value2) < 0)
+                throw new ArithmeticException("Number cannot be negative.");
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UIntBase.cs]: .NET throws `System.OverflowException` whether it is overflow or underflow with description "Arithmetic operation resulted in an overflow."

So if substraction cause underflow then i expect same behavior of .NET

File: Stratis.SmartContracts.Tests/UInt256Tests.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            Assert.Equal(x, new UInt256(x.ToBytes()));
+        }
+
+        [Fact]
+        [Trait("UnitTest", "UnitTest")]
+        public void CanAdd()
+        {
+            UInt256 v1 = new UInt256("0000000000000000000000000000000000000000000000000000000000000325");
+            UInt256 v2 = new UInt256("0000000000000000000000000000000000000000000000000000000000000F0F");
+            UInt256 v3 = v1 + v2;
```
Comment (YakupIpek) [inline Stratis.SmartContracts.Tests/UInt256Tests.cs:74]: Add another test for proving that overflow case throws `System.OverFlowException` for `+` operator.

File: Stratis.SmartContracts.Tests/UInt256Tests.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            Assert.Equal(new UInt256("0000000000000000000000000000000000000000000000000000000000001234"), v3);
+        }
+
+        [Fact]
+        [Trait("UnitTest", "UnitTest")]
+        public void CanSubtract()
+        {
+            UInt256 v1 = new UInt256("0000000000000000000000000000000000000000000000000000000000001234");
+            UInt256 v2 = new UInt256("0000000000000000000000000000000000000000000000000000000000000325");
+            UInt256 v3 = v1 - v2;
```
Comment (YakupIpek) [inline Stratis.SmartContracts.Tests/UInt256Tests.cs:84]: Add another test for proving that underflow case throws `System.OverFlowException` for `-` operator.

File: Stratis.SmartContracts.Tests/UInt256Tests.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            Assert.Equal(new UInt256("0000000000000000000000000000000000000000000000000000000000000F0F"), v3);
+        }
+
+        [Fact]
+        [Trait("UnitTest", "UnitTest")]
+        public void CanMultiply()
+        {
+            UInt256 v1 = new UInt256("0000000000000000000000000000000000000000000000000000000000001234");
+            UInt256 v2 = new UInt256("0000000000000000000000000000000000000000000000000000000000005678");
+            UInt256 v3 = v1 * v2;
```
Comment (YakupIpek) [inline Stratis.SmartContracts.Tests/UInt256Tests.cs:94]: Add another test for proving that overflow case throws `System.OverFlowException` for `*` operator.

**Review by YakupIpek — CHANGES_REQUESTED** [review]

File: Stratis.SmartContracts/UInt128.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Numerics;
+
+namespace Stratis.SmartContracts
+{
+    public class UInt128 : UIntBase
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt128.cs]: Types used for contracts are struct as design like Address type. It is also BigInteger is struct as well. I think there are good reason that it is intentionally designed as struct.

File: Stratis.SmartContracts/UInt256.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Numerics;
+
+namespace Stratis.SmartContracts
+{
+    public class UInt256 : UIntBase
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt256.cs]: Same as https://github.com/stratisproject/Stratis.SmartContracts/pull/6/files#r556582399

File: Stratis.SmartContracts/IInternalTransactionExecutor.cs [hunk 9]
```diff
@@ -12,19 +14,19 @@ public interface IInternalTransactionExecutor
         /// <param name="smartContractState">State representing existing contract's context.</param>
         /// <param name="addressTo">Where the funds will be transferred to.</param>
         /// <param name="amountToTransfer">The amount to send, in satoshi.</param>
-        ITransferResult Transfer(ISmartContractState smartContractState, Address addressTo, ulong amountToTransfer);
+        ITransferResult Transfer(ISmartContractState smartContractState, Address addressTo, Amount amountToTransfer);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/IInternalTransactionExecutor.cs]: I think there is misunderstanding here. It is not requested to support smaller units than satoshi for Cirrus token.

File: Stratis.SmartContracts/IInternalTransactionExecutor.cs [hunk 10]
```diff
@@ -12,19 +14,19 @@ public interface IInternalTransactionExecutor
...
 
         /// <summary>
         /// Calls a method on another contract.
         /// </summary>
-        ITransferResult Call(ISmartContractState smartContractState, Address addressTo, ulong amountToTransfer, string methodName, object[] parameters, ulong gasLimit = 0);
+        ITransferResult Call(ISmartContractState smartContractState, Address addressTo, Amount amountToTransfer, string methodName, object[] parameters, ulong gasLimit = 0);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/IInternalTransactionExecutor.cs]: I think there is misunderstanding here. It is not requested to support smaller units than satoshi for Cirrus token.

File: Stratis.SmartContracts/IInternalTransactionExecutor.cs [hunk 11]
```diff
@@ -12,19 +14,19 @@ public interface IInternalTransactionExecutor
...
 
         /// <summary>
         /// Creates a new contract.
         /// </summary>
         /// <typeparam name="T">Type of contract to create.</typeparam>
         /// <param name="smartContractState">State repository to track and persist changes to the contract.</param>
         /// <param name="amountToTransfer">Amount to send, in satoshi.</param>
-        ICreateResult Create<T>(ISmartContractState smartContractState, ulong amountToTransfer, object[] parameters, ulong gasLimit = 0);
+        ICreateResult Create<T>(ISmartContractState smartContractState, Amount amountToTransfer, object[] parameters, ulong gasLimit = 0);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/IInternalTransactionExecutor.cs]: Same as others

File: Stratis.SmartContracts/SmartContract.cs [hunk 12]
```diff
@@ -1,4 +1,5 @@
 ﻿using System.Globalization;
+using Amount = Stratis.SmartContracts.UInt256;
```
Comment (YakupIpek) [inline Stratis.SmartContracts/SmartContract.cs]: What is the reason of given it an alias ?

File: Stratis.SmartContracts.Tests/UInt256Tests.cs (hunk 1 above)
Comment (quantumagi) [reply Stratis.SmartContracts.Tests/UInt256Tests.cs]: Not sure how we could simulate checked behavior. Is it really necessary or desirable to suppress these errors?

The internal `BigInteger` can handle any size number, so we can't rely on it for what you are describing.

**Review by YakupIpek — CHANGES_REQUESTED** [review]

File: Stratis.SmartContracts.Tests/UInt256Tests.cs (hunk 1 above)
Comment (YakupIpek) [reply Stratis.SmartContracts.Tests/UInt256Tests.cs]: These details should to be documented that these types will reject overflows cases with Exception whether it is in checked state or unchecked state

File: Stratis.SmartContracts/SmartContract.cs (hunk 12 above)
Comment (quantumagi) [reply Stratis.SmartContracts/SmartContract.cs]: The code reads a little bit better and if one day we need to change it again we just change what the alias maps to.

**Comment by quantumagi** [comment]
@mrtpain, I suppose we may need to update `IInternalTransactionExecutor` as well?

File: Stratis.SmartContracts/IMessage.cs [hunk 13]
```diff
@@ -15,6 +15,6 @@ public interface IMessage
         /// <summary>
         /// The number of coins sent with this call (in satoshis).
         /// </summary>
-        ulong Value { get; }
+        UInt256 Value { get; }
```
Comment (YakupIpek) [inline Stratis.SmartContracts/IMessage.cs]: Cirrus token values should not updated.

File: Stratis.SmartContracts/IPersistentState.cs [hunk 14]
```diff
@@ -64,6 +64,12 @@ public interface IPersistentState
         /// </summary>
         ulong GetUInt64(string key);
 
+        /// <summary>
+        /// Gets the bytes set at the value pointed to by the given key, and uses <see cref="ISerializer.ToUInt256"/> to
+        /// deserialize the value to a uint256. If deserialization is unsuccessful, returns default(uint256).
+        /// </summary>
+        UInt256 GetUInt256(string key);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/IPersistentState.cs:77]: We shouldn't add methods for UInt128 as well ?

File: Stratis.SmartContracts/IPersistentState.cs [hunk 15]
```diff
@@ -134,6 +140,12 @@ public interface IPersistentState
         /// </summary>
         void SetUInt64(string key, ulong value);
 
+        /// <summary>
+        /// Uses <see cref="ISerializer.Serialize(ulong)"/> to serialize a uint256 to its 32-byte representation.
+        /// Sets the serialized bytes against the given key in state storage.
+        /// </summary>
+        void SetUInt256(string key, UInt256 value);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/IPersistentState.cs:159]: We shouldn't add methods for UInt128 as well ?

File: Stratis.SmartContracts/IPersistentState.cs [hunk 16]
```diff
@@ -134,6 +140,12 @@ public interface IPersistentState
         /// </summary>
         void SetUInt64(string key, ulong value);
 
+        /// <summary>
+        /// Uses <see cref="ISerializer.Serialize(ulong)"/> to serialize a uint256 to its 32-byte representation.
```
Comment (YakupIpek) [inline Stratis.SmartContracts/IPersistentState.cs]: Comments needs update

File: Stratis.SmartContracts/Message.cs [hunk 17]
```diff
@@ -9,9 +9,9 @@ public sealed class Message : IMessage
         public Address Sender { get; }
 
         /// <inheritdoc/>
-        public ulong Value { get; }
+        public UInt256 Value { get; }
```
Comment (YakupIpek) [inline Stratis.SmartContracts/Message.cs]: Cirrus token values should not updated.

File: Stratis.SmartContracts/Message.cs [hunk 18]
```diff
@@ -9,9 +9,9 @@ public sealed class Message : IMessage
...
 
-        public Message(Address contractAddress, Address sender, ulong value)
+        public Message(Address contractAddress, Address sender, UInt256 value)
```
Comment (YakupIpek) [inline Stratis.SmartContracts/Message.cs]: Cirrus token values should not updated.

File: Stratis.SmartContracts/UInt128.cs [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+{
+    public struct UInt128
+    {
+        const int WIDTH = 16;
+
+        private UIntBase value;
+
+        public static UInt256 Zero = 0;
+        public static UInt256 MinValue = 0;
+        public static UInt256 MaxValue = new UInt256((BigInteger.One << (WIDTH * 8)) - 1);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt128.cs]: These types should to be UInt128

File: Stratis.SmartContracts/UIntBase.cs [hunk 20]
```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Numerics;
+
+namespace Stratis.SmartContracts
+{
+    public struct UIntBase : IComparable
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UIntBase.cs]: I think this should stay as private

File: Stratis.SmartContracts.Tests/UInt256Tests.cs (hunk 1 above)
Comment (YakupIpek) [reply Stratis.SmartContracts.Tests/UInt256Tests.cs]: There are pros/cons of this behavior

pros 
-BigInteger is already .net built-in type and designed to handle overflow cases and we are not changing default behavior. Surely this is confusing bit of.
-It is actually safer that these numeric types doesn't let overflows. 


cons
-Checking overflow and throw exception is extra computation and gas cost and It may not needed for every case.
-It is more expected behavior that arithmetic operations behave same as other numeric types in .net regarding to checked and unchecked contexts
-Coders should to be aware these behaviors by reading documentation. 

I request to you to add documentation on methods and operators for(Add,Multiply and Substract) that they will throws OverflowException and developers can easily see it. 

@mrtpain Let me know your feedbacks.

File: Stratis.SmartContracts/UInt128.cs [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+
+        public static implicit operator UInt128(ulong value)
+        {
+            return new UInt128(value);
+        }
+
+        public static implicit operator UInt128(long value)
+        {
+            if (value < 0)
+                throw new ArgumentException("Only positive or zero values are allowed.", nameof(value));
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt128.cs]: Implicity or explicit conversations always throws OverflowException for numeric types and i think we should to follow same behavior.

File: Stratis.SmartContracts/UInt128.cs [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+
+            return new UInt128((ulong)value);
+        }
+
+        public static implicit operator UInt128(int value)
+        {
+            if (value < 0)
+                throw new ArgumentException("Only positive or zero values are allowed.", nameof(value));
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt128.cs]: Implicity or explicit conversations always throws OverflowException for numeric types and i think we should to follow same behavior.

File: Stratis.SmartContracts/UInt256.cs [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+
+        public static implicit operator UInt256(ulong value)
+        {
+            return new UInt256(value);
+        }
+
+        public static implicit operator UInt256(long value)
+        {
+            if (value < 0)
+                throw new ArgumentException("Only positive or zero values are allowed.", nameof(value));
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt256.cs]: Implicity or explicit conversations always throws OverflowException for numeric types and i think we should to follow same behavior.

File: Stratis.SmartContracts/UInt256.cs [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+
+            return new UInt256((ulong)value);
+        }
+        
+        public static implicit operator UInt256(int value)
+        {
+            if (value < 0)
+                throw new ArgumentException("Only positive or zero values are allowed.", nameof(value));
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt256.cs]: Implicity or explicit conversations always throws OverflowException for numeric types and i think we should to follow same behavior.

File: Stratis.SmartContracts/ISerializer.cs [hunk 25]
```diff
@@ -91,6 +91,11 @@ public interface ISerializer
         /// </summary>
         ulong ToUInt64(byte[] val);
 
+        /// <summary>
+        /// Deserializes the first 32 bytes of a byte array into an uint256. If the given bytes are null, empty, or deserialization fails, returns default(uint256).
+        /// </summary>
+        UInt256 ToUInt256(byte[] val);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/ISerializer.cs:102]: Add Serialize method for UInt256 also add UInt128 methods for Serialize/Deserialize.

File: Stratis.SmartContracts/ISerializer.cs [hunk 26]
```diff
@@ -91,6 +91,16 @@ public interface ISerializer
         /// </summary>
         ulong ToUInt64(byte[] val);
 
+        /// <summary>
+        /// Deserializes the first 16 bytes of a byte array into an UInt128. If the given bytes are null, empty, or deserialization fails, returns default(UInt128).
+        /// </summary>
+        UInt128 ToUInt128(byte[] val);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/ISerializer.cs:97]: Still `Serialize` methods missing for `UInt128` and `UInt256`

Comment (quantumagi) [reply Stratis.SmartContracts/ISerializer.cs:97]: Look here: https://github.com/stratisproject/StratisFullNode/pull/365/files?file-filters%5B%5D=#diff-de42f079363455aa4a9a840704eb10090823a7fb1a3984dce43511a5d87dbba6

File: Stratis.SmartContracts.Tests/UInt256Tests.cs (hunk 1 above)
Comment (mrtpain) [reply Stratis.SmartContracts.Tests/UInt256Tests.cs]: I don't know the differences in gas costs of using checked() or not, so my opinion isn't really backed by supporting arguments. I don't think throwing on overflow is problematic, probably actually preferred to prevent accidental overflows but again, I'm not sure the cost in gas in doing so. 

Overall, if I had to choose, I would prefer that an OverflowException be thrown automatically, rather than not having checked/unchecked options or a new equivalent. Either way, I agree these differences should be documented.

File: Stratis.SmartContracts/UInt128.cs [hunk 27]
```diff
@@ -0,0 +1,200 @@
...
+
+        public UInt128(long b)
+        {
+            this.value = new UIntBase(WIDTH, b);
+        }
+
+        public UInt128(ulong b)
+        {
+            this.value = new UIntBase(WIDTH, b);
+        }
```
Comment (mrtpain) [inline Stratis.SmartContracts/UInt128.cs:44]: Should int and uint be supported?

File: Stratis.SmartContracts/UInt128.cs [hunk 28]
```diff
@@ -0,0 +1,200 @@
...
+
+        public static implicit operator UInt128(long value)
+        {
+            return new UInt128(value);
+        }
+
+        public static implicit operator UInt128(int value)
+        {
+            return new UInt128(value);
+        }
```
Comment (mrtpain) [inline Stratis.SmartContracts/UInt128.cs:168]: Should we add `uint`?

File: Stratis.SmartContracts/UInt128.cs [hunk 29]
```diff
@@ -0,0 +1,200 @@
...
+
+        public static implicit operator ulong(UInt128 value)
+        {
+            return (ulong)value.value.GetValue();
+        }
```
Comment (mrtpain) [inline Stratis.SmartContracts/UInt128.cs]: Would there be any benefit or consequences to allowing casting to and from all numeric types? int, uint, long, ulong, uint128, uint256? I know some of the lower types might never or rarely be used but would it be worth supporting just incase?

**Comment by mrtpain** [comment]
> @mrtpain, I suppose we may need to update `IInternalTransactionExecutor` as well?

I missed this comment yesterday, I think since that's dealing with CRS values, ulong is fine

File: Stratis.SmartContracts/ISerializer.cs (hunk 26 above)
Comment (YakupIpek) [reply Stratis.SmartContracts/ISerializer.cs:97]: I look at there but i didn't understand still :). Those methods aren't necessary in here ?

Comment (quantumagi) [reply Stratis.SmartContracts/ISerializer.cs:97]: Ok I see what you mean now. We have to add that in the next PR.
